util/fipstools/acvp/modulewrapper/modulewrapper.cc

```diff
@@ -3072,6 +3132,171 @@ static bool ED25519SigVer(const Span<const uint8_t> args[],
   return write_reply({Span<const uint8_t>(reply)});
 }
 
+template <int nid>
+static bool ML_DSA_KEYGEN(const Span<const uint8_t> args[],
+                          ReplyCallback write_reply) {
+  const Span<const uint8_t> seed = args[0];
+
+  //init params of the correct size based on provided nid
+  ml_dsa_params params;
+  if (nid == NID_MLDSA44) {
+    ml_dsa_44_params_init(&params);
+  }
+  else if (nid == NID_MLDSA65) {
+    ml_dsa_65_params_init(&params);
+  }
+  else if (nid == NID_MLDSA87) {
+    ml_dsa_87_params_init(&params);
+  }
```

Comment:
Why isn't this using the EVP API? This low level API isn't the approved one right?

Reply:
You can't use the approved entrance point because ACVP needs to be able to specify the entropy/seed data.

Reply:
Sean is right. We don't want to provide an EVP API that allows customers to insert their own seeds, but we need to provide keygen and sign with a random seed so we can KAT to a known result. FIPS 204 fixes this for us, by providing internal functions that we have implemented that take randonmess for testing purposes. The external variants call these same internal functions when you want real randomness, as they provide the seed at run time and pass to internal.

For ML-KEM we were able to use the EVP API because I added extra EVP functions that take seeds as input, as for ML-KEM this can actually be useful when in a protocol (e.g., MLS).

Reply:
See Section 5 of [FIPS 204](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.204.pdf)

> The signing, verifying, and key generation functions can be split into “external” and “internal” components
> to simplify APIs and Cryptographic Algorithm Validation Program (CAVP) testing. The external components
> generate randomness and perform various checks before calling their internal counterparts. The internal
> components are deterministic and can assume that the external components did not encounter error
> conditions